pandas/core/sorting.py

```diff
@@ -499,7 +508,10 @@ def ensure_key_mapped(values, key: Optional[Callable], levels=None):
             result = Index(result)
         else:
             type_of_values = type(values)
-            result = type_of_values(result)  # try to revert to original type otherwise
+            # GH 35922. Support sorting objects that raise when cast to their type
+            if not isinstance(result, type_of_values):
+                # try to revert to original type otherwise
+                result = type_of_values(result)
```

Comment:
Some of the supported types don't like being cast to themselves.

Reply:
what does this fail on?

Reply:
Seems like other fixes made this safeguard unnecessary. Now that we pass a `Series` into the function, there are no more failures. If I remember correctly, it was failing on tz-aware datetimes, when a key function would strip the timezone information, and then the cast would fail, because it wouldn't know what to do with the timezones in `values`.